components/ws-manager/pkg/manager/pod_controller.go

```diff
@@ -46,6 +46,10 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 		Type:   watch.Modified,
 		Object: &pod,
 	})
+	r.Monitor.eventpool.Add(queue, watch.Event{
+		Type:   watch.Deleted,
+		Object: &pod,
+	})
```

Comment:
should this go to line 36, i.e. only add a `Deleted` event when the pod isn't found?

do we care about reliability of catching these delete events? They won't always be received iirc, e.g. it's possible for multiple events to get batched together, or the reconciler finding the Pod cached in the client on its last received event

Reply:
> In that case, it is impossible to determine if the pod was a workspace, as in line 40.

Hmm I see. We can configure an `EventFilter` when setting up the controller with the manager to filter out Pod events for non-workspace pods, but that still doesn't give us the workspace ID from the annotations. We could keep track of a cache of pod name to workspace ID and use that during delete events?

(https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/reconcile/reconcile.go#L87 has some more info as well)

> I'm not familiar with Reconcile fo k8s but it seems that MaxConcurrentReconciles is 1.
https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.0/pkg/controller#Options

yeah there won't be concurrent reconcile events, but multiple events for one resource can still get batched together and sent as one event to the controller. See e.g. https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/doc.go#L45 with a better explanation

Reply:
The update events and existing code is fine, it's the new delete events that in the current implementation will get added to the monitor queue on every Pod update as well. Haven't looked into the monitor code too much, but wouldn't it be a problem if the monitor received a delete event for a Pod even though the pod only got updated and still exists?

Happy to sync as well 👍

Reply:
So, we have the full workspace pod information on the [Deleted](https://github.com/kubernetes/apimachinery/blob/fd8a60496be5e4ce19d586738adb48ac6fa66ef9/pkg/watch/watch.go#L62) event.
Also, we have existing code for handling the [Deleted event](https://github.com/gitpod-io/gitpod/blob/a362f98d33fdc83cfe050bc1a910f0ec998ddd3c/components/ws-manager/pkg/manager/monitor.go#L188) comes in. But we don't know if the behavior works correctly for now.

Or, we could do by once the Deleted event comes in, call the [onChange](https://github.com/gitpod-io/gitpod/blob/38fc38c4ef06e56dfecab8a36931cc6cbdf8e41b/components/ws-manager/pkg/manager/manager.go#L1339-L1366) function to update the current status to all the clients and return.

Note: we must ensure that deleting the workspace pod would not trigger two events (Modified + Deleted).